src/utils/latin.ts

```diff
@@ -0,0 +1,67 @@
+/* eslint-disable prefer-spread */
+
+/**
+ * This function is an optimization for small basic latin strings.
+ * @internal
+ * @remarks
+ * ### Important characteristics:
+ * - If the uint8array or distance between start and end is 0 this function returns an empty string
+ * - If the byteLength of the string is 1, 2, or 3 we invoke String.fromCharCode and manually offset into the buffer
+ * - If the byteLength of the string is less than or equal to 20 an array of bytes is built and `String.fromCharCode.apply` is called with the result
+ * - If any byte exceeds 128 this function returns null
+ *
+ * @param uint8array - A sequence of bytes that may contain basic latin characters
+ * @param start - The start index from which to search the uint8array
+ * @param end - The index to stop searching the uint8array
+ * @returns string if all bytes are within the basic latin range, otherwise null
+ */
+export function tryLatin(uint8array: Uint8Array, start: number, end: number): string | null {
+  if (uint8array.length === 0) {
+    return '';
+  }
+
+  const stringByteLength = end - start;
+  if (stringByteLength === 0) {
+    return '';
+  }
+
+  if (stringByteLength === 1 && uint8array[start] < 128) {
+    return String.fromCharCode(uint8array[start]);
+  }
+
+  if (stringByteLength === 2 && uint8array[start] < 128 && uint8array[start + 1] < 128) {
+    return String.fromCharCode(uint8array[start]) + String.fromCharCode(uint8array[start + 1]);
+  }
+
+  if (
+    stringByteLength === 3 &&
+    uint8array[start] < 128 &&
+    uint8array[start + 1] < 128 &&
+    uint8array[start + 2] < 128
+  ) {
+    return (
+      String.fromCharCode(uint8array[start]) +
+      String.fromCharCode(uint8array[start + 1]) +
+      String.fromCharCode(uint8array[start + 2])
+    );
+  }
+
+  if (stringByteLength <= 20) {
+    let basicLatin = true;
+    const latinBytes = [];
+    for (let i = start; i < end; i++) {
+      const byte = uint8array[i];
+      if (byte > 127) {
+        basicLatin = false;
+        break;
+      }
+      latinBytes.push(byte);
+    }
+
+    if (basicLatin) {
+      return String.fromCharCode.apply(String, latinBytes);
```

Comment:
Just a question, why are we calling `apply` here and not just `fromCharCode` directly?

Reply:
Good question, I realize I hadn't actually tested, I assumed that spreading would be worse for performance than apply, figured that spreading might make a copy of the array. I tested and found the opposite, looks like spreading is ever so slightly better, I don't think it makes much difference in performance but it is certainly nicer on the eyes.

```txt
stringLen: 7
toUTF8_try_latin x 13,747,586 ops/sec ±1.06% (92 runs sampled)
toUTF8_try_latin_spread x 14,469,017 ops/sec ±0.29% (93 runs sampled)

stringLen: 9
toUTF8_try_latin x 12,610,412 ops/sec ±0.45% (93 runs sampled)
toUTF8_try_latin_spread x 13,128,721 ops/sec ±0.25% (96 runs sampled)

stringLen: 16
toUTF8_try_latin x 9,209,619 ops/sec ±0.18% (95 runs sampled)
toUTF8_try_latin_spread x 9,485,863 ops/sec ±0.20% (96 runs sampled)
```